librz/main/rz-diff.c

```diff
@@ -1,1305 +1,1454 @@
-			for (i = 0; i < op->a_len; i++) {
-				printf("%02x", op->a_buf[i]);
-			}
+			rz_diff_error_opt(ctx, DIFF_OPT_ERROR, "option -d argument '%s' is not a recognized algorithm.\n", algorithm);
 		}
-		if (ro->disasm) {
-			int i;
-			printf("+++ 0x%08" PFMT64x "  ", op->b_off);
-			if (!ro->core) {
-				ro->core = opencore(ro, NULL);
-			}
-			for (i = 0; i < op->b_len; i++) {
-				printf("%02x", op->b_buf[i]);
+	} else if (type) {
+		rz_diff_ctx_set_opt(ctx, DIFF_OPT_UNIFIED);
+
+		if (!strcmp(type, "bytes")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_BYTES);
+		} else if (!strcmp(type, "lines")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_LINES);
+		} else if (!strcmp(type, "functions")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_FUNCTIONS);
+		} else if (!strcmp(type, "classes")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_CLASSES);
+		} else if (!strcmp(type, "command")) {
+			if (!ctx->command_a) {
+				rz_diff_error_opt(ctx, DIFF_OPT_ERROR, "option -t '%s' requires -0 <command>.\n", type);
 			}
-			printf("\n");
-			if (ro->core) {
-				int len = RZ_MAX(4, op->b_len);
-				RzAsmCode *ac = rz_asm_mdisassemble(ro->core->rasm, op->b_buf, len);
-				char *acbufasm = strdup(ac->assembly);
-				if (ro->quiet) {
-					char *bufasm = rz_str_prefix_all(acbufasm, "+ ");
-					printf("%s\n", bufasm);
-					free(bufasm);
-					free(acbufasm);
-				} else {
-					char *bufasm = rz_str_prefix_all(acbufasm, Color_GREEN "+ ");
-					printf("%s\n" Color_RESET, bufasm);
-					free(bufasm);
-					free(acbufasm);
-				}
-				rz_asm_code_free(ac);
+			if (!ctx->command_b) {
+				ctx->command_b = ctx->command_a;
 			}
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_COMMAND);
+		} else if (!strcmp(type, "entries")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_ENTRIES);
+		} else if (!strcmp(type, "fields")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_FIELDS);
+		} else if (!strcmp(type, "imports")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_IMPORTS);
+		} else if (!strcmp(type, "libraries")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_LIBRARIES);
+		} else if (!strcmp(type, "sections")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_SECTIONS);
+		} else if (!strcmp(type, "strings")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_STRINGS);
+		} else if (!strcmp(type, "symbols")) {
+			rz_diff_ctx_set_type(ctx, DIFF_TYPE_SYMBOLS);
 		} else {
-			printf(" => ");
-			for (i = 0; i < op->b_len; i++) {
-				printf("%02x", op->b_buf[i]);
-			}
-			printf(" 0x%08" PFMT64x "\n", op->b_off);
+			rz_diff_error_opt(ctx, DIFF_OPT_ERROR, "option -t argument '%s' is not a recognized type.\n", type);
 		}
-		return 1;
+	} else {
+		rz_diff_error_opt(ctx, DIFF_OPT_USAGE, "option -t or -d is required to be specified.\n");
 	}
-	return 0;
 }
 
-void print_bytes(const void *p, size_t len, bool big_endian) {
-	size_t i;
-	for (i = 0; i < len; i++) {
-		ut8 ch = ((ut8 *)p)[big_endian ? (len - i - 1) : i];
-		if (write(1, &ch, 1) != 1) {
-			break;
-		}
+/* This is terrible because can eat a lot of memory */
```

Comment:
Nice to know. Maybe add `FIXME:` so it will be easier to spot/find?

Reply:
actually this is limited by a check on the file size

Reply:
`MEGABYTE(100)` is max buffer for this